### ISSUE TITLE
hashCode collision for ValueObjects

### ISSUE THREAD

**Issue by xpr0ger** [javers/javers#543]
I have faced with a rare issue related to value objects comparison. Maybe because of a huge amount of nested object, at some point method `System.identityHashCode` generates repeated hash codes, that is leading to quite unpredictable behavior. The problem appears in method `NodeReuser.reverseCdoIdMapKey` the reason is `System.identityHashCode` are no uniqueness guarantied.

Could you give me some advice, if it is possible, how I can avoid this problem without altering the source code of Javers?

**Comment by bartoszwalacik** [comment]
thanks for reporting, if the hashCode collision occurs, we should fix Javers code. Could you isolate and reproduce the problem in a test case?

**Comment by xpr0ger** [comment]
Ok. I'll try to do it.

**Comment by xpr0ger** [comment]
I tried to reproduce the issue within stable conditions, but no success. It always appears with different data sets, I have no any idea how to reproduce it in the tests. Bellow, I put the cause of this issue. The consequences is a false-positive difference of two entities, sometimes.

I'll try again when I have more free time, but for now, maybe it could help you.
```java
Map<Integer, NestedEntity> hashMap = new HashMap<>();
for (Long i = 0L; i < Long.MAX_VALUE; i++) {
    NestedEntity tmpEntity = new NestedEntity(i);
    Integer hashCode = System.identityHashCode(tmpEntity);
    if (null != hashMap.get(hashCode)) {
        NestedEntity collision = hashMap.get(hashCode);
        this.logger.debug("Objects with id " + collision.getId() + " has same hash code as " + tmpEntity.getId());
        return null;
    }

    hashMap.put(hashCode, tmpEntity);
}
```

**Comment by bartoszwalacik** [comment]
Thanks, do you have an idea where in Javers' code we use bare System.identityHashCode to compare objects?

**Comment by xpr0ger** [comment]
[NodeReuser.reverseCdoIdMapKey](https://github.com/javers/javers/blob/master/javers-core/src/main/java/org/javers/core/graph/NodeReuser.java) at line 83. The method is used the `saveForReuse` method. When an object with the same hashCode appears then it overwrites the old one.

**Comment by bartoszwalacik** [comment]
Ok, thanks for the hint, I will check it

**Comment by bartoszwalacik** [comment]
So the problematic method is:

```
private Object reverseCdoIdMapKey(Cdo cdo) {
        if (cdo.getGlobalId() instanceof InstanceId) {
            return cdo.getGlobalId();
        }
        return System.identityHashCode(cdo.getWrappedCdo().get());
}
```
I'm not sure why we treat Entities in a different way than ValueObjects

**Comment by xpr0ger** [comment]
First of all, I should have been mention of it before, the issue is related only to ValueObjects. Now about of `equals()`, I checked it yesterday, and yes it should use `equals()`. But, when we are running into collision it does not matter because of we will get two equal integers out of `identityHashCode` those keys we use as the keys for hashMap and those keys will be and same and equals at same time.

**Comment by bartoszwalacik** [comment]
You are right, bare hashcode as a key is not enough. We should create a wrapping object, with smarter equals()

**Comment by bartoszwalacik** [comment]
fix is merged 
https://github.com/javers/javers/pull/674
will be released in 3.9.6
